README.md
[![Dart](https://github.com/altera2015/linalg/actions/workflows/dart.yml/badge.svg)](https://github.com/altera2015/linalg/actions/workflows/dart.yml) [![Pub](https://img.shields.io/pub/v/linalg.svg)](https://pub.dartlang.org/packages/linalg)

# linalg

A Simple Linear Algebra Package. 

This package is intended to be a portable, easy to use linear algebra package. The library does not have any dependencies outside of Dart itself, thus making it portable and easy to integrate.
Internally all numbers are stored as Dart Doubles ( 64 bit as specified in the IEEE 754 standard ).

Our goal is to keep the code readable, documented and maintainable.

## Short Example

Just a quick example on how to do matrix multiplication.

```dart
final Matrix a = Matrix([[1, 2], [3, 4]]);
final Vector b = Vector.column([2, 3]);
final Matrix e = Matrix([[8], [18]]);
Matrix result = a * b;
print(result);
print(result == e);
```
This prints
```dart
[[8.0], [18.0]]
true
```

## Complete Example

A more extensive example with various matrix operations. See the [Matrix API](https://pub.dartlang.org/documentation/linalg/latest/linalg.matrix/Matrix-class.html) and [Vector API](https://pub.dartlang.org/documentation/linalg/latest/linalg.vector/Vector-class.html) for the full details.

```dart
import 'package:linalg/linalg.dart';

void example() {
	// *****************************
	// Lets solve a linear equation.
	// *****************************
	//
	// A * B = E
	//
	// We have B and E, we have to find A.
	//
	// A * B * B' = E * B'
	//
	// Were B' = inverse of B.
	//
	// A * I = E * B'
	//
	// Where I = identity matrix
	//
	// A = E * B'
	final Matrix B = Matrix([[2, 0], [1, 2]]);
	final Matrix E = Matrix([[4, 4], [10, 8]]);


	Matrix Bi = B.inverse();
	Matrix A_calc = E * Bi;

	final Matrix A = Matrix([[1, 2], [3, 4]]);
	print("The calculated A_calc = $A_calc, the expected A is $A, they are ${A_calc==A?'':'not'} the same.");
	// Expected: The calculated A = [[1.0, 2.0], [3.0, 4.0]], the expected A is [[1.0, 2.0], [3.0, 4.0]], they are  the same.



	// *****************************
	// Lets do some more matrix math
	// *****************************
	//
	// Next let multiply Matrix A by 3.
	Matrix Am = A * 3.0;
	print(Am);
	// Expecting: [[3.0, 6.0], [9.0, 12.0]]

	// Now add matrix B to A
	Matrix AmPlusB = Am + B;
	print(AmPlusB);
	// Expecting: [[5.0, 6.0], [10.0, 14.0]]

	// What is the determinant of the A matrix?
	print("The determinant of A = ${A.det()}");
	// Expecting The determinant of A = -2.0
}
```

## Installation

Add linalg as a dependency to your flutter project.

```yaml
dependencies:
  linalg: ^0.4.0
```

and at the top of your dart file add:

```dart
import 'package:linalg/linalg.dart';
```

## Attribution

Original code came from:
https://pub.dartlang.org/packages/toaster_linear

We ended up rewriting considerable parts of the code and 
adding tests and documentation.


CHANGELOG.md
## [0.4.0] -- 7/8/2021

* Added null safety.

## [0.3.2] -- 5/9/2019

* Updated Coverage collection to work with recent dart builds.

## [0.3.1] -- 12/24/2018

* Fixed Vector.toList and Vector.toMatrix to return deepCopies by default. Added optional parameters
  to return references where possible for optimization use.

## [0.3.0] -- 12/24/2018

* Added matrix.columnVector, Matrix.rowVector and Matrix.copy
* Changed toVector to create a deep copy by default, but allowed it to reuse Matrix storage for performance.

## [0.2.0] -- 12/15/2018

* Added Vector.sum, Vector.mean, Vector.norm, Vector.manhattanNorm and Vector.toList with tests and docs.
* Deprecated Vector.distance for the better name Vector.norm.

## [0.1.1] -- 11/27/2018

* Documentation updates
* Added Matrix.trace, Vector.crossProduct and Vector.dotProduct

## [0.1.0] -- 11/26/2018

* Slight documentation modifications.

## [0.0.6] -- 11/26/2018

* Added Vector class.

## [0.0.5] -- 11/21/2018

* Dart minimum back to 2.0.0 fixed linter errors.

## [0.0.4] -- 11/21/2018

* Increased Dart minimum to 2.1.0 to allow int to double assignments.

## [0.0.3] -- 11/21/2018

* Removed flutter dependency

## [0.0.2] - 11/21/2018

* Added examples
* Added Travis CI
* Improved readme

## [0.0.1] - 11/21/2018.

* Initial release



test/test_all.dart
library linalg.test;

import 'matrix_test.dart' as matrixTest;
import 'vector_test.dart' as vectorTest;

void main() {
  matrixTest.main();
  vectorTest.main();
}


test/linalg_test.dart
library linalg.test.linalg;

import 'package:test/test.dart';
import 'package:linalg/linalg.dart';

void main() {

  test("Identity Matrix generator", () {
    final Matrix eye = Matrix.eye(3);
    expect(eye, Matrix([[1.0,0.0,0.0],[0.0,1.0,0.0],[0.0,0.0,1.0]]));
  });

  test('Matrix value access tests', () {
    final Matrix mt = Matrix( [ [1.0, 3.0 ], [5.0,-2.0] ]);
    expect(mt[0][0], 1.0);
    expect(mt[0][1], 3.0);
    expect(mt[1][0], 5.0);
    expect(mt[1][1], -2.0);
  });

  test('Matrix fill test', () {
    final Matrix m = Matrix.fill(2,2,10.0);
    expect(m[0][0], 10.0);
    expect(m[0][1], 10.0);
    expect(m[1][0], 10.0);
    expect(m[1][1], 10.0);
  });

  test('Matrix Cut test', () {

    final Matrix m = Matrix( [
      [1.0, 2.0, 3.0],
      [4.0, 5.0, 6.0],
      [7.0, 8.0, 9.0],
    ]);

    final Matrix expected = Matrix( [
      [1.0, 3.0],
      [7.0, 9.0]
    ]);

    expect(m.cut(1,1), expected);
  });


  test("Matrix Addition", () {
    final Matrix a = Matrix.eye(3);
    final Matrix b = Matrix([[1.0, 1.0, 2.0],[3.0,1.0,3.0], [5.0,6.0,7.0]]);
    expect(a+b, Matrix([[2.0,1.0,2.0],[3.0,2.0,3.0],[5.0,6.0,8.0]]));
  });

  test("Matrix Subtraction", () {
    final Matrix a = Matrix([[1.0, 1.0, 2.0],[3.0,1.0,3.0], [5.0,6.0,7.0]]);
    final Matrix b = Matrix.eye(3);
    expect(a-b, Matrix([[0.0,1.0,2.0],[3.0,0.0,3.0],[5.0,6.0,6.0]]));
  });

  test("Matrix Transpose", () {
    final Matrix a = Matrix([[1.0,2.0],[3.0,4.0],[5.0,6.0]]);
    expect(a.transpose(), Matrix([[1.0, 3.0, 5.0],[2.0,4.0,6.0]]));
  });

  test("Matrix Negate", () {
    final Matrix a = Matrix.eye(3);
    expect(~a, Matrix([[-1.0, 0.0, 0.0],[0.0,-1.0,0.0],[0.0,0.0,-1.0]]));
  });

  test('2x2 Matrix Determinant test', () {

    expect(Matrix( [ [1.0, 3.0 ],  [5.0, -2.0]]).det(), -17);
    expect(Matrix( [ [2.0, 1.0 ],  [4.0,  2.0]]).det(), 0);
    expect(Matrix( [ [4.0, -1.0 ], [-3.0, 2.0]]).det(), 5);
    expect(Matrix( [ [4.0, -3.0 ], [1.0,  2.0]]).det(), 11);

  });

  test('3x3 Matrix Determinant test', () {

    expect(Matrix( [ [1.0, 3.0, 2.0 ],  [4.0,1.0,3.0],  [2.0,2.0,0.0  ]]).det(), 24);
    expect(Matrix( [ [1.0, 0.0, 2.0 ],  [1.0,3.0,4.0],  [0.0,6.0,0.0  ]]).det(), -12);
    expect(Matrix( [ [3.0, -2.0, 4.0 ], [2.0,-4.0,5.0], [1.0,8.0,2.0  ]]).det(), -66);
    expect(Matrix( [ [8.0, -1.0, 9.0 ], [3.0,1.0,8.0],  [11.0,0.0,17.0]]).det(), 0);

  });

  test('4x4 Matrix Determinant test', () {
    expect(Matrix( [ [1.0, 2.0, 1.0, 0.0 ], [0.0,3.0,1.0, 1.0], [-1.0,0.0,3.0,1.0], [3.0,1.0,2.0,0.0]]).det(), 16);
  });

  test('4x4 Matrix Determinant test', () {
    expect(Matrix( [ [1.0, 2.0, 1.0, 0.0 ], [0.0,3.0,1.0, 1.0], [-1.0,0.0,3.0,1.0], [3.0,1.0,2.0,0.0]]).det(), 16);
  });

  test('Co Factors Matrix test', () {
    expect(Matrix([ [3.0, 0.0, 2.0], [2.0, 0.0, -2.0], [0.0, 1.0, 1.0]]).coFactors(), Matrix([ [2.0, -2.0, 2.0], [2.0, 3.0, -3.0], [0.0, 10.0, 0.0]]));
  });

  test("Matrix Multiplication test", () {
    final Matrix a1 = Matrix([[1.0, 2.0], [3.0, 4.0]]);
    final Matrix b1 = Matrix([[2.0, 0.0], [1.0, 2.0]]);
    final Matrix e1 = Matrix([[4.0, 4.0], [10.0, 8.0]]);
    expect(a1 * b1, e1);

    final Matrix a2 = Matrix([[1.0, 2.0], [3.0, 4.0], [1.0, 2.0]]);
    final Matrix b2 = Matrix([[2.0, 0.0], [1.0, 2.0]]);
    final Matrix e2 = Matrix([[4.0, 4.0], [10.0, 8.0], [4.0, 4.0]]);
    expect(a2 * b2, e2);
  });

  test("Matrix inversion 2x2", () {
    final Matrix a = Matrix([[4.0,7.0],[2.0,6.0]]);
    expect(a.inverse(), Matrix([[0.6, -0.7],[-0.2,0.4]]));
  });

  test("Matrix inversion 3x3", () {
    final Matrix a = Matrix([[1.0,3.0, 3.0],[1.0, 4.0, 3.0], [1.0,3.0,4.0]]);
    expect(a.inverse(), Matrix([[7.0, -3.0, -3.0],[-1.0, 1.0, 0.0],[-1.0, 0.0, 1.0]]));
  });

  test("Matrix inversion 4x4", () {
    final Matrix a = Matrix([[1.0,2.0, 4.0,1.0],[5.0, 3.0, 1.0,2.0], [4.0,2.0,4.0,1.0], [6.0, 3.0, 2.0, 1.0]]);
    expect(a.inverse(), Matrix([[-1/3, 0.0, 1/3,0.0],[20/27, -2/9, -29/27,7/9],[1/27,-1/9, 8/27, -1/9],[-8/27, 8/9, 17/27, -10/9]]));
  });

  // Test exceptions

  test("Exception: Creating matrix with invalid dimensions",(){

    expect(()=>Matrix.eye(-2), throwsA( TypeMatcher<MatrixInvalidDimensions>()) );

    final Matrix a = Matrix.eye(3);
    final Matrix b = Matrix.eye(2);
    expect(()=>a*b, throwsA( TypeMatcher<MatrixInvalidDimensions>()) );

  });

  test("Exception: Multiplying matrix with invalid argument",(){

    final Matrix a = Matrix.eye(3);
    final String b = "this makes no sense";
    expect(()=>a*b, throwsA( TypeMatcher<MatrixUnsupportedOperation>()) );

  });

  test("Exception: Determinant of non-square matrix",(){

    final Matrix a = Matrix([[1.0, 2.0], [3.0, 4.0], [1.0, 2.0]]);
    expect(()=>a.det(), throwsA( TypeMatcher<MatrixInvalidDimensions>()) );

  });

  test("Exception: Cofactors of non-square matrix",(){

    final Matrix a = Matrix([[1.0, 2.0], [3.0, 4.0], [1.0, 2.0]]);
    expect(()=>a.coFactors(), throwsA( TypeMatcher<MatrixInvalidDimensions>()) );

  });

  test("Exception: Determinant of 1x1 matrix",(){

    final Matrix a = Matrix.eye(1);
    expect(()=>a.det(), throwsA( TypeMatcher<MatrixInvalidDimensions>()) );

  });

  test("Exception: Cofactor of 2x2 matrix",(){

    final Matrix a = Matrix.eye(2);
    expect(()=>a.coFactors(), throwsA( TypeMatcher<MatrixInvalidDimensions>()) );

  });

  test("Exception: Impossible Inversion", () {
    final Matrix a = Matrix([[1.0,3.0, 3.0],[2.0, 6.0, 6.0], [1.0,3.0,4.0]]);
    expect(()=>a.inverse(), throwsA( TypeMatcher<MatrixNoInverseException>()) );
  });

}


test/matrix_test.dart
library linalg.test.matrix;

import 'package:linalg/linalg.dart';
import 'package:test/test.dart';

void main() {
  group("Matrix", () {
    test("Identity Matrix generator", () {
      final Matrix eye = Matrix.eye(3);
      expect(
          eye,
          Matrix([
            [1.0, 0.0, 0.0],
            [0.0, 1.0, 0.0],
            [0.0, 0.0, 1.0]
          ]));
    });

    test('value access tests', () {
      final Matrix mt = Matrix([
        [1.0, 3.0],
        [5.0, -2.0]
      ]);
      expect(mt[0][0], 1.0);
      expect(mt[0][1], 3.0);
      expect(mt[1][0], 5.0);
      expect(mt[1][1], -2.0);
    });

    test('fill test', () {
      final Matrix m = Matrix.fill(2, 2, 10.0);
      expect(m[0][0], 10.0);
      expect(m[0][1], 10.0);
      expect(m[1][0], 10.0);
      expect(m[1][1], 10.0);
    });

    test('cut test', () {
      final Matrix m = Matrix([
        [1.0, 2.0, 3.0],
        [4.0, 5.0, 6.0],
        [7.0, 8.0, 9.0],
      ]);

      final Matrix expected = Matrix([
        [1.0, 3.0],
        [7.0, 9.0]
      ]);

      expect(m.cut(1, 1), expected);
    });

    test("addition", () {
      final Matrix a = Matrix.eye(3);
      final Matrix b = Matrix([
        [1.0, 1.0, 2.0],
        [3.0, 1.0, 3.0],
        [5.0, 6.0, 7.0]
      ]);
      expect(
          a + b,
          Matrix([
            [2.0, 1.0, 2.0],
            [3.0, 2.0, 3.0],
            [5.0, 6.0, 8.0]
          ]));
    });

    test("subtraction", () {
      final Matrix a = Matrix([
        [1.0, 1.0, 2.0],
        [3.0, 1.0, 3.0],
        [5.0, 6.0, 7.0]
      ]);
      final Matrix b = Matrix.eye(3);
      expect(
          a - b,
          Matrix([
            [0.0, 1.0, 2.0],
            [3.0, 0.0, 3.0],
            [5.0, 6.0, 6.0]
          ]));
    });

    test("transpose", () {
      final Matrix a = Matrix([
        [1.0, 2.0],
        [3.0, 4.0],
        [5.0, 6.0]
      ]);
      expect(
          a.transpose(),
          Matrix([
            [1.0, 3.0, 5.0],
            [2.0, 4.0, 6.0]
          ]));
    });

    test("negate", () {
      final Matrix a = Matrix.eye(3);
      expect(
          ~a,
          Matrix([
            [-1.0, 0.0, 0.0],
            [0.0, -1.0, 0.0],
            [0.0, 0.0, -1.0]
          ]));
    });

    test('2x2 Matrix Determinant test', () {
      expect(
          Matrix([
            [1.0, 3.0],
            [5.0, -2.0]
          ]).det(),
          -17);
      expect(
          Matrix([
            [2.0, 1.0],
            [4.0, 2.0]
          ]).det(),
          0);
      expect(
          Matrix([
            [4.0, -1.0],
            [-3.0, 2.0]
          ]).det(),
          5);
      expect(
          Matrix([
            [4.0, -3.0],
            [1.0, 2.0]
          ]).det(),
          11);
    });

    test('3x3 Matrix Determinant test', () {
      expect(
          Matrix([
            [1.0, 3.0, 2.0],
            [4.0, 1.0, 3.0],
            [2.0, 2.0, 0.0]
          ]).det(),
          24);
      expect(
          Matrix([
            [1.0, 0.0, 2.0],
            [1.0, 3.0, 4.0],
            [0.0, 6.0, 0.0]
          ]).det(),
          -12);
      expect(
          Matrix([
            [3.0, -2.0, 4.0],
            [2.0, -4.0, 5.0],
            [1.0, 8.0, 2.0]
          ]).det(),
          -66);
      expect(
          Matrix([
            [8.0, -1.0, 9.0],
            [3.0, 1.0, 8.0],
            [11.0, 0.0, 17.0]
          ]).det(),
          0);
    });

    test('4x4 Matrix Determinant test', () {
      expect(
          Matrix([
            [1.0, 2.0, 1.0, 0.0],
            [0.0, 3.0, 1.0, 1.0],
            [-1.0, 0.0, 3.0, 1.0],
            [3.0, 1.0, 2.0, 0.0]
          ]).det(),
          16);
    });

    test('4x4 Matrix Determinant test', () {
      expect(
          Matrix([
            [1.0, 2.0, 1.0, 0.0],
            [0.0, 3.0, 1.0, 1.0],
            [-1.0, 0.0, 3.0, 1.0],
            [3.0, 1.0, 2.0, 0.0]
          ]).det(),
          16);
    });

    test('Co Factors Matrix test', () {
      expect(
          Matrix([
            [3.0, 0.0, 2.0],
            [2.0, 0.0, -2.0],
            [0.0, 1.0, 1.0]
          ]).coFactors(),
          Matrix([
            [2.0, -2.0, 2.0],
            [2.0, 3.0, -3.0],
            [0.0, 10.0, 0.0]
          ]));
    });

    test("multiplication test", () {
      final Matrix a1 = Matrix([
        [1.0, 2.0],
        [3.0, 4.0]
      ]);
      final Matrix b1 = Matrix([
        [2.0, 0.0],
        [1.0, 2.0]
      ]);
      final Matrix e1 = Matrix([
        [4.0, 4.0],
        [10.0, 8.0]
      ]);
      expect(a1 * b1, e1);

      final Matrix a2 = Matrix([
        [1.0, 2.0],
        [3.0, 4.0],
        [1.0, 2.0]
      ]);
      final Matrix b2 = Matrix([
        [2.0, 0.0],
        [1.0, 2.0]
      ]);
      final Matrix e2 = Matrix([
        [4.0, 4.0],
        [10.0, 8.0],
        [4.0, 4.0]
      ]);
      expect(a2 * b2, e2);
    });

    test("inversion 2x2", () {
      final Matrix a = Matrix([
        [4.0, 7.0],
        [2.0, 6.0]
      ]);
      expect(
          a.inverse(),
          Matrix([
            [0.6, -0.7],
            [-0.2, 0.4]
          ]));
    });

    test("inversion 3x3", () {
      final Matrix a = Matrix([
        [1.0, 3.0, 3.0],
        [1.0, 4.0, 3.0],
        [1.0, 3.0, 4.0]
      ]);
      expect(
          a.inverse(),
          Matrix([
            [7.0, -3.0, -3.0],
            [-1.0, 1.0, 0.0],
            [-1.0, 0.0, 1.0]
          ]));
    });

    test("inversion 4x4", () {
      final Matrix a = Matrix([
        [1.0, 2.0, 4.0, 1.0],
        [5.0, 3.0, 1.0, 2.0],
        [4.0, 2.0, 4.0, 1.0],
        [6.0, 3.0, 2.0, 1.0]
      ]);
      expect(
          a.inverse(),
          Matrix([
            [-1 / 3, 0.0, 1 / 3, 0.0],
            [20 / 27, -2 / 9, -29 / 27, 7 / 9],
            [1 / 27, -1 / 9, 8 / 27, -1 / 9],
            [-8 / 27, 8 / 9, 17 / 27, -10 / 9]
          ]));
    });

    test('Matrix trace', () {
      expect(
          Matrix([
            [1.0, 2.0, 1.0, 0.0],
            [0.0, 3.0, 1.0, 1.0],
            [-1.0, 0.0, 3.0, 1.0],
            [3.0, 1.0, 2.0, 0.0]
          ]).trace(),
          7.0);
    });

    test('columnVector', () {
      Matrix m = Matrix([
        [1.0, 2.0, 3.0, 1.0],
        [4.0, 5.0, 6.0, 2.0],
        [7.0, 8.0, 9.0, 3.0],
      ]);

      Vector vc = m.columnVector(0);
      expect(vc, Vector.column([1.0, 4.0, 7.0]));

      // columnVector is always a deep copy
      m[0][0] = 100.0;
      expect(vc, Vector.column([1.0, 4.0, 7.0]));
    });

    test('toVector deepCopy', () {
      Matrix m = Matrix([
        [1.0, 2.0, 3.0, 1.0],
      ]);

      Vector vc = m.toVector();
      expect(vc, Vector.row([1.0, 2.0, 3.0, 1.0]));

      // rowVector by default is a deep copy.
      m[0][0] = 100.0;
      expect(vc, Vector.row([1.0, 2.0, 3.0, 1.0]));
    });

    test('toVector reference', () {
      Matrix m = Matrix([
        [1.0, 2.0, 3.0, 1.0],
      ]);

      Vector vc = m.toVector(false);
      expect(vc, Vector.row([1.0, 2.0, 3.0, 1.0]));

      // rowVector by default is a deep copy.
      m[0][0] = 100.0;
      expect(vc, Vector.row([100.0, 2.0, 3.0, 1.0]));
    });

    test('rowVector deepCopy', () {
      Matrix m = Matrix([
        [1.0, 2.0, 3.0, 1.0],
        [4.0, 5.0, 6.0, 2.0],
        [7.0, 8.0, 9.0, 3.0],
      ]);

      Vector vc = m.rowVector(0);
      expect(vc, Vector.row([1.0, 2.0, 3.0, 1.0]));

      // rowVector by default is a deep copy.
      m[0][0] = 100.0;
      expect(vc, Vector.row([1.0, 2.0, 3.0, 1.0]));
    });

    test('rowVector reference', () {
      Matrix m = Matrix([
        [1.0, 2.0, 3.0, 1.0],
        [4.0, 5.0, 6.0, 2.0],
        [7.0, 8.0, 9.0, 3.0],
      ]);

      Vector vc = m.rowVector(0, false);
      expect(vc, Vector.row([1.0, 2.0, 3.0, 1.0]));

      // rowVector with forces shared storage.
      m[0][0] = 100.0;
      expect(vc, Vector.row([100.0, 2.0, 3.0, 1.0]));
    });

    // Test exceptions

    test("Exception: Creating matrix with invalid dimensions", () {
      expect(() => Matrix.eye(-2), throwsA(TypeMatcher<MatrixInvalidDimensions>()));

      final Matrix a = Matrix.eye(3);
      final Matrix b = Matrix.eye(2);
      expect(() => a * b, throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });

    test("Exception: Multiplying matrix with invalid argument", () {
      final Matrix a = Matrix.eye(3);
      final String b = "this makes no sense";
      expect(() => a * b, throwsA(TypeMatcher<MatrixUnsupportedOperation>()));
    });

    test("Exception: Determinant of non-square matrix", () {
      final Matrix a = Matrix([
        [1.0, 2.0],
        [3.0, 4.0],
        [1.0, 2.0]
      ]);
      expect(() => a.det(), throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });

    test("Exception: Cofactors of non-square matrix", () {
      final Matrix a = Matrix([
        [1.0, 2.0],
        [3.0, 4.0],
        [1.0, 2.0]
      ]);
      expect(() => a.coFactors(), throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });

    test("Exception: Determinant of 1x1 matrix", () {
      final Matrix a = Matrix.eye(1);
      expect(() => a.det(), throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });

    test("Exception: Cofactor of 2x2 matrix", () {
      final Matrix a = Matrix.eye(2);
      expect(() => a.coFactors(), throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });

    test("Exception: Impossible Inversion", () {
      final Matrix a = Matrix([
        [1.0, 3.0, 3.0],
        [2.0, 6.0, 6.0],
        [1.0, 3.0, 4.0]
      ]);
      expect(() => a.inverse(), throwsA(TypeMatcher<MatrixNoInverseException>()));
    });

    test("Exception: Matrix to Vector that is not a vector", () {
      final Matrix a = Matrix.eye(2);
      expect(() => a.toVector(), throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });

    test("Exception: Trace on non-square matrix", () {
      final Matrix a = Matrix([
        [1.0, 3.0, 3.0],
        [2.0, 6.0, 6.0]
      ]);
      expect(() => a.trace(), throwsA(TypeMatcher<MatrixInvalidDimensions>()));
    });
  });
}


example/example.dart
import 'package:linalg/linalg.dart';

void readMeExample() {
  final Matrix a = Matrix([
    [1, 2],
    [3, 4]
  ]);
  final Vector b = Vector.column([2, 3]);
  final Matrix e = Matrix([
    [8],
    [18]
  ]);
  Matrix result = a * b;
  print(result);
  print(result == e);
}

/// Example code for using the Matrix class.
void main() {

  readMeExample();

  // *****************************
  // Lets solve a linear equation.
  // *****************************
  //
  // A * B = E
  //
  // We have B and E, we have to find A.
  //
  // A * B * B' = E * B'
  //
  // Were B' = inverse of B.
  //
  // A * I = E * B'
  //
  // Where I = identity matrix
  //
  // A = E * B'
  final Matrix B = Matrix([
    [2.0, 0.0],
    [1.0, 2.0]
  ]);
  final Matrix E = Matrix([
    [4.0, 4.0],
    [10.0, 8.0]
  ]);

  Matrix Bi = B.inverse();
  Matrix A_calc = E * Bi;

  final Matrix A = Matrix([
    [1.0, 2.0],
    [3.0, 4.0]
  ]);
  print(
      "The calculated A = $A_calc, the expected A is $A, they are ${A_calc == A ? '' : 'not'} the same.");
  // Expected: The calculated A = [[1.0, 2.0], [3.0, 4.0]], the expected A is [[1.0, 2.0], [3.0, 4.0]], they are  the same.

  // *****************************
  // Lets do some more matrix math
  // *****************************
  //
  // Next let multiply Matrix A by 3.
  Matrix Am = A * 3.0;
  print(Am);
  // Expecting: [[3.0, 6.0], [9.0, 12.0]]

  // Now add matrix B to A
  Matrix AmPlusB = Am + B;
  print(AmPlusB);
  // Expecting: [[5.0, 6.0], [10.0, 14.0]]

  // What is the determinant of the A matrix?
  print("The determinant of A = ${A.det()}");
  // Expecting The determinant of A = -2.0
}
